sklearn/metrics/_classification.py

```diff
@@ -2688,3 +2747,94 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
             raise
     y_true = np.array(y_true == pos_label, int)
     return np.average((y_true - y_prob) ** 2, weights=sample_weight)
+
+
+def multiclass_brier_score_loss(y_true, y_prob, sample_weight=None, labels=None):
+    r"""Compute the Brier score loss.
+
+    The smaller the Brier score loss, the better, hence the naming with "loss".
+    The Brier score measures the mean squared difference between the predicted
+    probability and the actual outcome.
+
+    For :math:`N` samples with :math:`C` different classes, the multi-class
+    Brier score is defined as:
+
+    .. math::
+        \frac{1}{N}\sum_{i=1}^{N}\sum_{c=1}^{C}(y_{ic} - \hat{y}_{ic})^{2}
+
+    where :math:`y_{ic}` is 1 if observation `i` belongs to class `c`,
+    otherwise 0 and :math:`\hat{y}_{ic}` is the predicted probability of
+    observation `i` for class `c`. The probabilities for `c` classes for
+    observation `i` should sum to 1.
+
+    The Brier score always takes on a value between [0, 2]. For the
+    binary case however, there is a more common definition of Brier score
+    implemented in :func:`brier_score_loss` that is exactly half of the value
+    returned by this function, thereby having a range between [0, 1].
```

Comment:
I think `multiclass_brier_score_loss` would be the first metric that is an explicit "multi-class" version of an existing metric. Was the reason for the new function because of the different ranges?

Reply:
Yes. See the discussion in https://github.com/scikit-learn/scikit-learn/pull/18699#issuecomment-718795716.

Reply:
I would definitely prefer to have one Brier score/loss and have it handle both (statistical/probabilistic) binary and multiclass classification.

By the general definition summing over classes (dropping sample index i) `sum_{c=1}^{C}(y_{c} - \hat{y}_{c})^{2}` one gets for the binary case `2 (y_{0} - \hat{y}_{0})^{2}`. So it is just a matter of a global factor (2 for the binary case). We could, for instance, devide by the number of classes, i.e. sum_c => 1/c sum_c.

Reply:
I would also be ok with an additional argument like `devide_by_classes` and, if necessary, deprecation cycles. But I have a strong feeling about having one Brier score only.

Reply:
On top, the brier score with the 1/class normalization is equal to `mean_squared_error(y_true, y_pred, multioutput='uniform_average',)` if we generate `y_pred=LabelBinarizer()`, i.e. treating multiclass as multioutput target.
And I would expect such a test.

Reply:
I am not so sure about the the `divide_by_classes` option. It's not something that is often done in the literature as far as I know for `n_classes > 2`. So doing it by default would be quite confusing. I prefer to be explicit with error messages that point from one method to the other and avoid any implicit misunderstanding.

Also, a nice side effect of having a separate function is that it's natural to chose between the two alternatives using by passing `scoring="neg_multiclass_brier_score"` or `scoring="neg_brier_score"` to a model selection tool without having to rely on a `partial`  / `make_scorer` call.

Reply:
🤔 Let's talk about this.

Reply:
In [roc_auc_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score),  we have a `multi_class` parameter with default set to `'raise'` and in the scorers we define unique strings for multiclass problems:

https://github.com/scikit-learn/scikit-learn/blob/42a34e81a2efd64ddd7b40b433765e7c361cb51e/sklearn/metrics/_scorer.py#L762-L764

For `brier_score`, we can adopt the same pattern, `multi_class=False` by default, which raises an error for multiclass problems. `multi_class=True` would switch to the new implementation. We can still have a string for scorers such as: `neg_brier_score_multi_class`.

Reply:
A score that by default raises an error for multiclass problems seems not very user friendly to me. I don't intend to get us stuck here, but those are my true concerns.

Reply:
> I am not so sure about the the divide_by_classes option. It's not something that is often done in the literature as far as I know for n_classes > 2.

As info: The original reference of [Brier "Verification of Forecasts expressed in Terms of Probability" (1950)](https://doi.org/10.1175/1520-0493(1950)078%3C0001:VOFEIT%3E2.0.CO;2) uses the bare sum over classes (`1/n` is for sample size), while the later reference [Brier "A New Vector Partition of the Probability Score" (1973)](https://doi.org/10.1175/1520-0450(1973)012<0595:ANVPOT>2.0.CO;2) uses the average over classes, i.e. `1/K = 1/n_classes`.

How about trusting the author's insight of 23 years of additional experience, use the `1/n_classes` normalisation as default and be consistent with the binary brier score we already have?